File: src/frontend/src/lib/scanner-config.js
/**
 * Scanner config field definitions.
 *
 * Each entry describes one backend config field so the UI can auto-generate
 * the settings form.  To add a new field, add one object here — no template
 * changes needed.
 */

export const SCANNER_CONFIG_FIELDS = [
  // ── Template Matching ──────────────────────────────────────────────
  {
    key: "confidence_threshold",
    label: "Konfidenz-Schwellwert",
    group: "Template Matching",
    type: "number",
    min: 0,
    max: 1,
    step: 0.05,
  },
  {
    key: "matching_method",
    label: "Matching-Methode",
    group: "Template Matching",
    type: "select",
    options: [
      { value: "TM_CCOEFF_NORMED", label: "Kreuzkorrelationskoeffizient (Standard)" },
      { value: "TM_CCORR_NORMED", label: "Kreuzkorrelation" },
      { value: "TM_SQDIFF_NORMED", label: "Quadratische Differenz" },
    ],
  },
  {
    key: "auto_verify_confidence",
    label: "Auto-Verifizierung ab",
    group: "Template Matching",
    type: "number",
    min: 0,
    max: 1,
    step: 0.05,
  },

  // ── Multi-Scale ────────────────────────────────────────────────────
  {
    key: "multi_scale_enabled",
    label: "Multi-Scale-Suche",
    group: "Multi-Scale",
    type: "toggle",
  },
  {
    key: "multi_scale_range",
    label: "Suchbereich (+/-)",
    group: "Multi-Scale",
    type: "number",
    min: 0.01,
    max: 0.5,
    step: 0.01,
  },
  {
    key: "multi_scale_steps",
    label: "Stufen",
    group: "Multi-Scale",
    type: "number",
    min: 1,
    max: 20,
    step: 1,
  },

  // ── Edge-Based Matching ────────────────────────────────────────────
  {
    key: "edge_matching_enabled",
    label: "Kanten-Matching",
    group: "Kanten-Matching",
    type: "toggle",
  },
  {
    key: "canny_low",
    label: "Canny unterer Schwellwert",
    group: "Kanten-Matching",
    type: "number",
    min: 0,
    max: 500,
    step: 10,
  },
  {
    key: "canny_high",
    label: "Canny oberer Schwellwert",
    group: "Kanten-Matching",
    type: "number",
    min: 0,
    max: 500,
    step: 10,
  },

  // ── Staff Removal ──────────────────────────────────────────────────
  {
    key: "staff_removal_before_matching",
    label: "Notenlinien vor Matching entfernen",
    group: "Notenlinien-Entfernung",
    type: "toggle",
  },

  // ── Dewarp (bent staff lines) ────────────────────────────────────────
  {
    key: "dewarp_enabled",
    label: "Krümmungskorrektur",
    group: "Krümmungskorrektur",
    type: "toggle",
  },
  {
    key: "dewarp_smoothing",
    label: "Glättung (px)",
    group: "Krümmungskorrektur",
    type: "number",
    min: 5,
    max: 200,
    step: 5,
  },

  // ── Masked Matching ────────────────────────────────────────────────
  {
    key: "masked_matching_enabled",
    label: "Maskiertes Matching",
    group: "Maskiertes Matching",
    type: "toggle",
  },
  {
    key: "mask_threshold",
    label: "Masken-Schwellwert",
    group: "Maskiertes Matching",
    type: "number",
    min: 0,
    max: 255,
    step: 5,
  },

  // ── NMS ────────────────────────────────────────────────────────────
  {
    key: "nms_iou_threshold",
    label: "NMS IoU-Schwellwert",
    group: "Non-Maximum Suppression",
    type: "number",
    min: 0,
    max: 1,
    step: 0.05,
  },
  {
    key: "nms_method",
    label: "NMS-Methode",
    group: "Non-Maximum Suppression",
    type: "select",
    options: [
      { value: "standard", label: "Standard (IoU)" },
      { value: "dilate", label: "Dilate (Proximity)" },
    ],
  },

  // ── Preprocessing ──────────────────────────────────────────────────
  {
    key: "adaptive_threshold_block_size",
    label: "Adaptiver Schwellwert Blockgr.",
    group: "Vorverarbeitung",
    type: "number",
    min: 3,
    max: 99,
    step: 2,
  },
  {
    key: "adaptive_threshold_c",
    label: "Adaptiver Schwellwert Konstante",
    group: "Vorverarbeitung",
    type: "number",
    min: 0,
    max: 50,
    step: 1,
  },
  {
    key: "morphology_kernel_size",
    label: "Morphologie-Kerngr.",
    group: "Vorverarbeitung",
    type: "number",
    min: 1,
    max: 10,
    step: 1,
  },
  {
    key: "deskew_method",
    label: "Deskew-Methode",
    group: "Vorverarbeitung",
    type: "select",
    options: [
      { value: "none", label: "Keine Korrektur" },
      { value: "hough", label: "Hough-Linien (schnell)" },
      { value: "projection", label: "Projektions-Optimierung (genau)" },
    ],
  },
];

/**
 * Group fields by their `group` property, preserving definition order.
 */
export function groupedFields() {
  const groups = [];
  const seen = new Map();
  for (const field of SCANNER_CONFIG_FIELDS) {
    if (!seen.has(field.group)) {
      const entry = { label: field.group, fields: [] };
      seen.set(field.group, entry);
      groups.push(entry);
    }
    seen.get(field.group).fields.push(field);
  }
  return groups;
}

File: src/backend/mv_hofki/models/scanner_config.py
"""ScannerConfig ORM model — single-row global pipeline settings."""

from __future__ import annotations

from sqlalchemy import Boolean, Float, Integer, String
from sqlalchemy.orm import Mapped, mapped_column

from mv_hofki.db.base import Base


class ScannerConfig(Base):
    __tablename__ = "scanner_config"

    id: Mapped[int] = mapped_column(primary_key=True)

    # Template matching
    confidence_threshold: Mapped[float] = mapped_column(
        Float, nullable=False, default=0.6
    )
    matching_method: Mapped[str] = mapped_column(
        String(30), nullable=False, default="TM_CCOEFF_NORMED"
    )

    # Multi-scale search
    multi_scale_enabled: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False
    )
    multi_scale_range: Mapped[float] = mapped_column(
        Float, nullable=False, default=0.05
    )
    multi_scale_steps: Mapped[int] = mapped_column(Integer, nullable=False, default=3)

    # Edge-based matching
    edge_matching_enabled: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False
    )
    canny_low: Mapped[int] = mapped_column(Integer, nullable=False, default=50)
    canny_high: Mapped[int] = mapped_column(Integer, nullable=False, default=150)

    # Staff removal before matching
    staff_removal_before_matching: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False
    )
    staff_removal_thickness_pct: Mapped[int] = mapped_column(
        Integer, nullable=False, default=100
    )
    staff_removal_symbol_padding: Mapped[int] = mapped_column(
        Integer, nullable=False, default=0
    )

    # Masked template matching
    masked_matching_enabled: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False
    )
    mask_threshold: Mapped[int] = mapped_column(Integer, nullable=False, default=200)

    # NMS settings
    nms_iou_threshold: Mapped[float] = mapped_column(Float, nullable=False, default=0.3)
    nms_method: Mapped[str] = mapped_column(
        String(20), nullable=False, default="standard"
    )

    # Preprocessing
    adaptive_threshold_block_size: Mapped[int] = mapped_column(
        Integer, nullable=False, default=15
    )
    adaptive_threshold_c: Mapped[int] = mapped_column(
        Integer, nullable=False, default=10
    )
    morphology_kernel_size: Mapped[int] = mapped_column(
        Integer, nullable=False, default=2
    )

    # Dewarp (bent staff line correction)
    dewarp_enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    dewarp_smoothing: Mapped[int] = mapped_column(Integer, nullable=False, default=50)

    # Deskew
    deskew_method: Mapped[str] = mapped_column(
        String(20), nullable=False, default="projection"
    )

    # Auto-verify
    auto_verify_confidence: Mapped[float] = mapped_column(
        Float, nullable=False, default=0.85
    )
